test/unit/coupling/indexing/CellIndexTest.cpp

```diff
@@ -0,0 +1,208 @@
+#include "coupling/CouplingMDDefinitions.h"
+#include "coupling/indexing/IndexingService.h"
+#include <cppunit/TestFixture.h>
+#include <cppunit/extensions/HelperMacros.h>
+
+using namespace coupling::indexing;
+
+enum class Testmode { conversion, operators, loop };
+
+class CellIndexTest : public CppUnit::TestFixture {
+
+  CPPUNIT_TEST_SUITE(CellIndexTest);
+  CPPUNIT_TEST(testAllConversions);
+  CPPUNIT_TEST(testAllOperators);
+  CPPUNIT_TEST(testAllLoops);
+  CPPUNIT_TEST(testOperatorStream);
+  CPPUNIT_TEST_SUITE_END();
+
+public:
+  using Trait = IndexTrait;
+
+  using T00 = CellIndex<3>;
+  using T01 = CellIndex<3, Trait::vector>;
+  using T02 = CellIndex<3, Trait::local>;
+  using T03 = CellIndex<3, Trait::vector, Trait::local>;
+  using T04 = CellIndex<3, Trait::md2macro>;
+  using T05 = CellIndex<3, Trait::vector, Trait::md2macro>;
+  using T06 = CellIndex<3, Trait::local, Trait::md2macro>;
+  using T07 = CellIndex<3, Trait::vector, Trait::local, Trait::md2macro>;
+  using T08 = CellIndex<3, Trait::noGhost>;
+  using T09 = CellIndex<3, Trait::vector, Trait::noGhost>;
+  using T10 = CellIndex<3, Trait::local, Trait::noGhost>;
+  using T11 = CellIndex<3, Trait::vector, Trait::local, Trait::noGhost>;
+  using T12 = CellIndex<3, Trait::md2macro, Trait::noGhost>;
+  using T13 = CellIndex<3, Trait::vector, Trait::md2macro, Trait::noGhost>;
+  using T14 = CellIndex<3, Trait::local, Trait::md2macro, Trait::noGhost>;
+  using T15 = CellIndex<3, Trait::vector, Trait::local, Trait::md2macro, Trait::noGhost>;
+
+  void setUp() {
+    _rank = 0;
+    _size = 1;
+#if (COUPLING_MD_PARALLEL == COUPLING_MD_YES)
+    MPI_Comm_size(MPI_COMM_WORLD, &_size);
+    MPI_Comm_rank(MPI_COMM_WORLD, &_rank);
+#endif
+
+    tarch::la::Vector<3, unsigned int> numberProcesses{(unsigned int)_size, 1, 1};
+    if (_size == 4)
+      numberProcesses = {2, 2, 1};
+
+    IndexingService<3>::getInstance().init({12}, numberProcesses, coupling::paralleltopology::XYZ, 3, (unsigned int)_rank);
```

Comment:
Would it make sense that _size is always 4 ? So even if MaMiCo is not built in parallel MPI mode then even the sequential MaMiCo unit tests would test that it also can work in parallel ? Because you don't need MPI to test if it could work with more than one processes.

Reply:
Wow! I didn't think of that. Amazing idea. It required a few more changes as I have to loop of all the ranks, but I did that and it works fine. Implemented in bfd7b2fae4a